Feed exercise implementation

ichachy opened the pull request:
First PR.
Would appreciate your notes.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:
```diff
@@ -1,5 +1,24 @@
+        @Json(name = "isPremium") val isPremium: Boolean,
+        @Json(name = "templateCategories") val templateCategories: List<String>?,
+        @Json(name = "templateName") val templateName: String?,
+        @Json(name = "templateThumbnailURI") val templateThumbnailURI: String)
```
Please add an empty line at the end of the file (eof).

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:
```diff
@@ -1,5 +1,24 @@
  */
+@JsonClass(generateAdapter = true)
+data class TemplatesMetadataItem(
+        @Json(name = "configuration") val configuration: String?,
```
This field is mandatory so it shouldn't be nullable. Same for the other fields of this class. So please remove the exclamation marks.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt:
```diff
@@ -1,5 +1,45 @@
+
+    companion object{
+        private const val BASE_URL = "https://assets.swishvideoapp.com/Android/demo/"
+        private const val SUFFIX_URL = "feed.json"
```
We need to found a better name for it. `FEED_URL`? We should also mark that this url is relative to the base url.

ichachy replied:
I changes the name to FEED_URL and added a comment that it's relative to base url

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt:
```diff
@@ -1,5 +1,45 @@
+                feedApiService = retrofit.create(FeedApiService::class.java)
+            }
+            return feedApiService!!
+        }
```
This method should be thread-safe. Since it might be accessed from any thread.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt:
```diff
@@ -1,5 +1,45 @@
+            return feedApiService!!
+        }
+    }
+}
```
eof.

nissimLightricks commented on build.gradle:
```diff
@@ -1,11 +1,11 @@
 buildscript {
-    ext.kotlin_version = '1.3.72'
+    ext.kotlin_version = '1.4.31'
     repositories {
         google()
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.0.0'
+        classpath 'com.android.tools.build:gradle:4.1.1'
```
This is not related to implementing the network layer. So I would put it in a separate commit or even separate pr.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,37 @@
+    abstract fun getAll(): Flowable<List<FeedEntity>>
+
+    @Query("SELECT COUNT(id) FROM feed_table")
+    abstract fun getTableSize(): Single<Int>
```
We should found a better name for this method. `getFeedEntitiesCount`?

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,37 @@
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    protected abstract fun insertSync(feedEntitiesList: List<FeedEntity>)
+}
```
eof

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/database/Entities.kt:
```diff
@@ -1,5 +1,16 @@
  */
+@Entity(tableName = "feed_table")
+data class FeedEntity(
+        @PrimaryKey val id: String,
```
Define explicitly the name for the id like you did for the other fields/columns to be consistent.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/database/Entities.kt:
```diff
@@ -1,5 +1,16 @@
+        @PrimaryKey val id: String,
+        @ColumnInfo(name = "thumbnailUrl") val thumbnailUrl: String,
+        @ColumnInfo(name = "isPremium") val isPremium: Boolean
+)
```
eof.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt:
```diff
@@ -1,5 +1,35 @@
+    abstract fun feedDao(): FeedDao
+
+    companion object{
+        private const val NAME = "feed_database"
```
We should found a better name. Name of what?

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt:
```diff
@@ -1,5 +1,35 @@
+                }
+                return database!!
+            }
+        }
```
Before acquiring the lock we should check for null. So in case it is non null we can return it directly without acquiring the lock at all.

nissimLightricks replied:
In Kotlin there is a built in solution for singleton object https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/lazy.html

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
-    fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
+    private var disposable: Disposable? = null
```
Disposable of what? Please give it a more descriptive name.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
+    companion object{
+        private const val PREFIX_URL = "https://assets.swishvideoapp.com/" +
+                                            "Android/demo/catalog/thumbnails/"
+        private const val NETWORK_ERROR = "Network error has accrued"
```
typo: accrued -> occurred.

nissimLightricks replied:
In general user facing strings are not hard coded but rather saved in a strings.xml and localised per languages.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
+        val feedItems: List<FeedItem>,
+        val isLoading: Boolean)
+
+    companion object{
```
We usually place companion objects at the bottom of the class. Same for inner classes.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
+     * Observes the FeedRepository and updates the stateInternal in case of any change
+     */
+    private fun observeFeedRepository() {
+        feedRepository.getAll().subscribeOn(Schedulers.io())
```
We should dispose the subscription once the user go away from the feature. This could be done by adding this reference to a `CompositeDisposable` and disposing it on `ViewModel#onCleared` method of the view model.

ichachy replied:
Instead of using CompositeDisposable, I added another disposable variable. On onCleared I dispose both disposable variables of the ViewModel

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
-        isLoading.value = false
-        isEmpty.value = true
+        updateState { copy(isLoading = true) }
+        disposable?.dispose()
```
I would change the order of the two. By first disposing the previous task - so removing any effect it could has on the view model, then launching a new task. In this specific case there is no a real difference since the method and the callback are running on the main thread, however as a best practice I would change the order.

nissimLightricks commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
-        isEmpty.value = true
+        updateState { copy(isLoading = true) }
+        disposable?.dispose()
+        disposable = feedRepository.refresh().subscribeOn(Schedulers.io())
```
We save an instance of the disposable but we are never disposing it.

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt:
```diff
@@ -0,0 +1,119 @@
+class FeedDaoTest {
+
+    @get:Rule
+    val instantExecutorRule = InstantTaskExecutorRule() // TODO ask Nissim
```
Please ask.

nissimLightricks replied:
I am not sure why we have this rule in the test. We can remove it.

ichachy replied:
I Forgot to hide the TODO comment. Nir explained it to me last week.

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -1,29 +1,74 @@
+            feedEntities == DummyFeedApiService.expectedFeedEntities
+        }
+    }
+}
```
eof

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt:
```diff
@@ -0,0 +1,119 @@
+        testObserver.assertValue {
+            it != null
+        }
+    }
```
RxJava components never emit `null` so you can remove this test.

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt:
```diff
@@ -0,0 +1,119 @@
+
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+
+    companion object{
```
Can you please move the companion object to the end of the class.

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt:
```diff
@@ -0,0 +1,119 @@
+
+        private val list1 = listOf(feedEntity1, feedEntity2)
+        private val list2 = listOf(feedEntity3, feedEntity4)
+        private val emptyList = emptyList<FeedEntity>()
```
What is the intent of this field?

ichachy replied:
I added a test to see  that the Dao object can handle an empty list in deleteAllAndInset method

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Test
+    fun testGetTableSize_initialTableSize_shouldReturn0(){
```
Please remove the prefix `test` from all the method name. We already have this information that the method is a test with the `@Test` annotation.

nissimLightricks commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt:
```diff
@@ -0,0 +1,119 @@
+        }
+        testObserver.dispose()
+    }
+}
```
eof

nissimLightricks reviewed (changes requested):
Good job 👏. Round.

Commit:
Versions updates

Commit:
Implementation of Network util

Added implememtation of classes that are responsible of parsering a JSON source of data and
map it's data to a DTO object. The code requires Retrofit, Moshi, RxJava libraries

Commit:
Implementation of Database util

Added implementation of classes that are responsible of saving the app's data in a local database. The code requires Room and testing libraries. A test class for DAO object was added.

Commit:
Implementation of Repository

Added an implementation of Repository class. The repository will fetch data from network and save it in a local database. A test classes were added for the repository

Commit:
Implementation of UI

Added an implementation of UI Fragment and ViewModel classes.

nissimLightricks reviewed (approved):
Nice!
